### PR TITLE
JS: add the NoSQL $where as a sink for js/code-injection

### PR DESCRIPTION
**Pull request by esbena**
This PR is on top of <https://github.com/github/codeql/pull/3457>.

Addresses <https://github.com/github/codeql-javascript-team/issues/83>

The four NoSQL models that we have all use the same operator name for the code operator that is modeled in this PR, so the implementation looks a bit redundant.

This change will enable us to flag the following of <https://github.com/github/codeql/issues/3434>:

-   <https://github.com/bkimminich/juice-shop/blob/446718f47b2a49c7eb557e9152f7232eb459008e/routes/trackOrder.js#L15>
-   <https://github.com/bkimminich/juice-shop/blob/446718f47b2a49c7eb557e9152f7232eb459008e/routes/showProductReviews.js#L29>

### CODE REVIEW

**Comment by esbena** [comment]
The minor [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/NoSQLCodeInjection_1589317728765) is uninteresting.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll [hunk 1]
```diff
@@ -713,3 +731,109 @@ private module Mongoose {
...
+    result = DataFlow::moduleImport("marsdb")
+    or
+    exists(DataFlow::TypeTracker t2 | result = getADb(t2).track(t2, t))
+  }
+
+  /** Gets a data flow node referring to a MarsDB collection. */
+  private DataFlow::SourceNode getACollection(DataFlow::TypeTracker t) {
+    t.start() and
+    // new Collection(...)
+    result = getADb(DataFlow::TypeTracker::end()).getAPropertyRead("Collection").getAnInvocation()
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:810]:
```suggestion
    result = getADb(DataFlow::TypeTracker::end()).getAPropertyRead("Collection").getAnInstantiation()
```

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll [hunk 2]
```diff
@@ -713,3 +731,109 @@ private module Mongoose {
...
+  private DataFlow::SourceNode getACollection(DataFlow::TypeTracker t) {
+    t.start() and
+    // db[COLLECTIONNAME]
+    result = getADb(DataFlow::TypeTracker::end()).getAPropertyRead()
+    or
+    exists(DataFlow::TypeTracker t2 | result = getACollection(t2).track(t2, t))
+  }
+
+  module CollectionMethodSignatures {
+    predicate interpretsArgumentAsQuery(string m, int queryArgIdx) {
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:760]: Missing qldoc on this module and predicate. 

Everything else has qldoc, so it looks weird having this omission.

**Comment by esbena** [comment]
juice-shop is in the evaluation above, but it is so fast to analyze that we only can see misoptimizations, so I think we are good on that front.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll (hunk 1 above)
Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:810]: The `new` requirement is pretty much 50/50 in the libraries that we model, and I am not afraid of potential FPs for this over-approximation, I am more worried about potential FNs. So the only benefit we gain from matching the documentation exactly is a tiny bit of performance.

Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:810]: I will adjust the comment above instead.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:816]: I tried to see what happened if I ran `require("marsdb").Collection()`, and the result is a type-error as `Collection` is a class and must be called with `new`.  
So I still prefer that we use `getAnInstantiation`. 

But an adjustment of the comment is also OK.

Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: OK, changed. 
(Thanks for being thorough with a semantic check, but I think it is a bit much to verify all models like this)

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: > I think it is a bit much to verify all models like this 

:+1: 
I very rarely do such a thorough check, but I just got curious about this corner of the API.

**Comment by asgerf** [comment]
What was the reason for using code-injection and not sql-injection? Code injection usually means RCE, which AFAIU isn't really the case here.

The name of the sql-injection query is actually `Database query built from user-controlled sources` which I think is still suitable for this sink.

**Comment by esbena** [comment]
> What was the reason for using code-injection and not sql-injection? 

Two reasons

1) Technical

https://github.com/github/codeql-javascript-team/issues/83 : 

>> The NoSql injection query will ignore string-values, but the $where operator in MongoDB is effectively a string-sink for code injection

So we need another label and different sanitizers to support the `$where`-sink properly.

2) Severity

https://docs.mongodb.com/manual/reference/operator/query/where/#op._S_where: 

>> Use the $where operator to pass either a string containing a JavaScript expression

The consequence of a `$where` injection is RCE. Most of the NoSql injections that we flag are "just" `DoS` or `select too many rows` vulnerabilities..

**Comment by asgerf** [comment]
> The NoSql injection query will ignore string-values

My point was that we don't have a NoSQL injection query. We have a "database query built from user-controlled sources query" that also encompasses SQL injection, which includes string sources. But if it causes RCE then LGTM.

**Comment by esbena** [comment]
> My point was that we don't have a NoSQL injection query. We have a "database query built from user-controlled sources query" that also encompasses SQL injection, which includes string sources.

Ah, good point. The SQL injection strings are however supposed to be sanitized differently than the code content for `$where`, so implementationwise, it is still simpler to put this in `js/code-injection`.

>  But if it causes RCE then LGTM.

:+1:. Lets leave it here and see if customers get confused about the lack of `js/sql-injections` for `$where` examples.

**Pull request merged**